docs(upgrading): update version upgrade summary (#262)
btpworks/chronicle: docs/upgrading.md

# Upgrading

## To 0.7x

### Accept and Verify JSON Web Tokens

We made changes to API authentication. Access to the API is now authenticated, and
users can provide a URI to the keyset used to sign the JWT bearer tokens. Users can
also provide a JSON pointer into the JWT claims for the Chronicle identity. A Chronicle
`AuthId` with a verified external ID can then be found in the GraphQL context.

We also converted the `JWT` `AuthId` variant to be a structure of both a json-pointer
-derived scalar and JWT claims.

### Anonymous Identity Variant

By default, Chronicle allows unauthenticated access to the GraphQL API, classifying
such a user as "anonymous." Adds an identity variant emanating from GraphQL or CLI,

> not authorized via Auth0, allowing for the option to pass a transactor key as a CLI

Reviewer: No need to mention Auth0 specifically, deployments may be using other providers.

or, if not, to generate an ephemeral transaction signing key for the anonymous agent.

Chronicle's GraphQL API will accept anonymous access as default unless disallowed
with the new `--require-auth` option. Only with that option must a JWKS server be
specified for JWT verification.

### OPA Execution

We introduced new interfaces for OPA policy execution and loading. The `OpaExecutor`
and `PolicyLoader` interfaces provide a flexible and straightforward way to evaluate
OPA policies.

The `OpaExecutor` interface takes a Chronicle `AuthId` identity and an arbitrary
JSON context, which it evaluates against an OPA policy. The `WasmtimeOpaExecutor`
evaluates OPA policies compiled to .wasm for execution.

The `PolicyLoader` interface offers two implementations. The `SawtoothPolicyLoader`
implementation loads an OPA policy rule that is compiled to .wasm from a Sawtooth
address. The `CliPolicyLoader` implementation provides support for loading an OPA
policy .wasm file via CLI arguments for a file path and policy entrypoint.

> Chronicle integrates OPA execution at its GraphQL API request point and in its Transaction

Reviewer 2: endpoint

from API calls will result in changes in state. If they don't, they won't be submitted
to the Transaction Processor and return `Submission::AlreadyRecorded` along with
GraphQL data containing `context` and a `null` `tx_id`.

## In-Domain Documentation Capabilities

Chronicle now provides users with the ability to generate a Chronicle domain with
documentation by using a domain.yaml file and adding documentation comments to it.

## `async-sawtooth-sdk`

Chronicle now uses `async-sawtooth-sdk` in this update, which contains generic client
types used by `opa-tp-protocol`. The transition to this new SDK has led to several
improvements.

The `LedgerReader` and `LedgerWriter` client abstractions are now in common, with
specialization for transactions and events in the `chronicle-protocol` and `opa-tp-protocol`
modules. The simulated transaction processors now use the actual Chronicle and OPA
TransactionHandlers during tests and in `devmode`.

## `prov:wasAttributedTo`

Chronicle now supports the `wasAttributedTo` PROV-O relationship.

## Endpoint offering JSON by IRI

Now, running Chronicle with the `serve-api` command, users can visit

> <http://localhost:9982/data/chronicle:agent:my-agent-id> and see a JSON summary.

Reviewer: also for entity and activity

## To 0.6x

### OPA TP, Protocol, and Skeletal CLI

We introduced an OPA policy and key registration transaction processor in this update.
The OPA policy and key registration transaction processor share similarities in design
with the Chronicle TP. Tested at the protobuf level, this feature is designed to
operate independently of the rest of the Chronicle project.

## To 0.5x

### Move from SQLite to PostgreSQL

Chronicle now utilizes PostgreSQL instead of SQLite as its local database.
This change brings several benefits, such as increased scalability and better

> performance for large-scale data management.

Reviewer 2: As an aside this is noticeable with the ETL on the discover project that's using a fake ledger. Chronicle is blessed performance wise as it is almost certainly dependent on something a lot slower than it so is always going to look good.

### `AuthId` Identity Type

We introduced an identity union in this update, which offers a versatile and flexible
approach to identity management. This new feature can be created from any of the
following options:

- Chronicle root key, identified by the agent name "chronicle"

> - An agent along with its corresponding key pair
> - A pointer that refers to the external id of an agent and its key pair through
>   dereferencing a JWT structure

Reviewer: ```suggestion
- an agent along with its corresponding key pair
- a pointer into a JWT structure, for retrieving the external id to use for the agent and its key pair.
```
